commit 6d8911403d991dfda859f5059af6f7c3071a2600
Author: Kirill Kapranov <kirill.kapranov@compulab.co.il>
Date:   Mon Aug 13 19:48:10 2018 +0300

    spi: fix IDR collision on systems with both fixed and dynamic SPI bus numbers
    
    commit 1a4327fbf4554d5b78d75b19a13d40d6de220159 upstream.
    
    On systems where some controllers get a dynamic ID assigned and some have
    a fixed number (e.g. from ACPI tables), the current implementation might
    run into an IDR collision: in case of a fixed bus number is gotten by a
    driver (but not marked busy in IDR tree) and a driver with dynamic bus
    number gets the same ID and predictably fails.
    
    Fix this by means of checking-in fixed IDsin IDR as far as dynamic ones
    at the moment of the controller registration.
    
    Fixes: 9b61e302210e (spi: Pick spi bus number from Linux idr or spi alias)
    Signed-off-by: Kirill Kapranov <kirill.kapranov@compulab.co.il>
    Signed-off-by: Mark Brown <broonie@kernel.org>
    Cc: stable@vger.kernel.org
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/spi/spi.c b/drivers/spi/spi.c
index ec395a6baf9c..a00d006d4c3a 100644
--- a/drivers/spi/spi.c
+++ b/drivers/spi/spi.c
@@ -2170,6 +2170,15 @@ int spi_register_controller(struct spi_controller *ctlr)
 		if (WARN(id < 0, "couldn't get idr"))
 			return id;
 		ctlr->bus_num = id;
+	} else {
+		/* devices with a fixed bus num must check-in with the num */
+		mutex_lock(&board_lock);
+		id = idr_alloc(&spi_master_idr, ctlr, ctlr->bus_num,
+			ctlr->bus_num + 1, GFP_KERNEL);
+		mutex_unlock(&board_lock);
+		if (WARN(id < 0, "couldn't get idr"))
+			return id == -ENOSPC ? -EBUSY : id;
+		ctlr->bus_num = id;
 	}
 	INIT_LIST_HEAD(&ctlr->queue);
 	spin_lock_init(&ctlr->queue_lock);
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1045_6d8911403d991dfda859f5059af6f7c3071a2600.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1045_6d8911403d991dfda859f5059af6f7c3071a2600.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
